### PR TITLE
fix(optimizer): forbid correlated input ref predicate push down through share

### PR DESCRIPTION
**Pull request by chenzl25**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

- Resolve https://github.com/risingwavelabs/risingwave/issues/13570
- We need to forbid correlated input ref predicate push through share operator.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Review by st1page — APPROVED** [review]
LGTM, can we just move the sub query unnesting before the generating of the share operator?

**Comment by chenzl25** [comment]
> LGTM, can we just move the sub query unnesting before the generating of the share operator?

Actually, some share operators are generated in the binder/planner (e.g.  CTE, view).

File: src/frontend/src/optimizer/plan_node/mod.rs [hunk 1]
```diff
@@ -423,7 +423,14 @@ impl PlanRef {
                     .map(|mut c| Condition {
                         conjunctions: c
                             .conjunctions
-                            .extract_if(|e| e.count_nows() == 0 && e.is_pure())
+                            .extract_if(|e| {
+                                // If predicates contain now, impure or correlated input ref, don't push through share operator.
```
Comment (st1page) [inline src/frontend/src/optimizer/plan_node/mod.rs:427]:
```suggestion
                                // If predicates contain now, impure or correlated input ref, don't push through share operator.
                                // The predicate with now() function is regarded as a temporal filter predicate, which will be transformed to a temporal filter operator and can not do the OR operation with other predicates.
```

Comment (chenzl25) [reply src/frontend/src/optimizer/plan_node/mod.rs:427]: It just reminds me that the `OR` operation can also cause a nested loop join.

**Pull request merged**